### PR TITLE
advanced noam with decay and accum scheduler

### PR DESCRIPTION
**Pull request by vince62s**
This PR introduces 2 things:

1) noam with decay (decay_method="noamwd")
When training on a lot of GPUs, the number of steps may be reduced (eg 50k) and you may want to decay faster the classic noam curve to a lower value.

2) accum gradient scheduler
When training on several GPUs, if you want a good performance you need a very high value of accum_count (eg 24 or more) but if you start training with such a huge batch size, it decreases the performance.
accum_count can now take a list of values, eg '4 8 16' which corresponds to a list of steps at which accum_count changes '10000 20000 30000'

Enjoy.

### CODE REVIEW

**Comment by vince62s** [comment]
@francoishernandez @guillaumekln if you got some time to review, thanks.

File: onmt/trainer.py [hunk 1]
```diff
@@ -112,18 +118,26 @@ def __init__(self, model, train_loss, valid_loss, optim,
...
+            if self.accum_count_l[i] > 1:
+                assert self.trunc_size == 0, \
+                    """To enable accumulated gradients,
+                       you must disable target sequence truncating."""
 
         # Set model in training mode.
         self.model.train()
 
+    def _accum_count(self, step):
+        for i in range(len(self.accum_steps)):
```
Comment (francoishernandez) [inline onmt/trainer.py]: here instead of looping every time shouldn't we update the `accum_count_l` list at each change so that all we need to check is `step > accum_count_l[0]`?

File: onmt/trainer.py [hunk 2]
```diff
@@ -90,7 +92,9 @@ class Trainer(object):
 
     def __init__(self, model, train_loss, valid_loss, optim,
                  trunc_size=0, shard_size=32,
-                 norm_method="sents", grad_accum_count=1, n_gpu=1, gpu_rank=1,
+                 norm_method="sents", accum_count='1',
```
Comment (guillaumekln) [inline onmt/trainer.py]: Why is the default value a string?

File: onmt/opts.py [hunk 3]
```diff
@@ -385,11 +385,14 @@ def train_opts(parser):
     group.add('--normalization', '-normalization', default='sents',
               choices=["sents", "tokens"],
               help='Normalization method of the gradient.')
-    group.add('--accum_count', '-accum_count', type=int, default=1,
+    group.add('--accum_count', '-accum_count', type=int, nargs='*',
```
Comment (guillaumekln) [inline onmt/opts.py]: Looks like the code expects `nargs='+'` (same for `accum_steps`).
